File: nodejs/packages/periph/src/chips/memory/_24aa02uid.js
'use strict';

const ADDR_UID_BASE   = 0xFC;
const ADDR_MFR_CODE   = 0xFA;
const ADDR_DEV_CODE   = 0xFB;
const WRITE_CYCLE_MS  = 5;
const ACK_POLL_MAX    = 20;
const PAGE_SIZE       = 8;

function _sleep(ms) {
    const end = Date.now() + ms;
    while (Date.now() < end) {}
}

/**
 * 24AA02UID 2K I2C EEPROM with 32-bit unique serial number — minimal interface.
 *
 * Provides read access to the chip's unique 32-bit serial number and basic
 * single-byte read/write access to the user EEPROM. The chip has no
 * configuration registers; it is ready for use immediately after power-on.
 *
 * Memory layout:
 * - 0x00-0x7F — 128 bytes general-purpose user EEPROM (R/W)
 * - 0x80-0xF9 — reserved, read-only
 * - 0xFA      — manufacturer code (always 0x29, Microchip)
 * - 0xFB      — device code (always 0x41)
 * - 0xFC-0xFF — 32-bit unique serial number, MSB first
 *
 * Default configuration (no registers to configure):
 * - User EEPROM is read/written as raw bytes (no interpretation)
 * - writeByte() waits for the internal write cycle to finish via
 *   ACK polling before returning (max 5 ms)
 * - All addresses 0x80-0xFF are write-protected; writes are silently
 *   ignored by the chip. Drivers accept the call but the data is not
 *   retained.
 */
class EEPROM24AA02UIDMinimal {
    /**
     * @param {object} transport - Configured I2C transport pointing at the device (address 0x50).
     */
    constructor(transport) {
        this._transport = transport;
    }

    /**
     * Read the chip's factory-programmed 32-bit unique serial number.
     * @returns {Buffer} 4-byte UID, MSB first (0xFC, 0xFD, 0xFE, 0xFF).
     */
    readUid() {
        return this._transport.writeRead(Buffer.from([ADDR_UID_BASE]), 4);
    }

    /**
     * Read a single byte from user EEPROM at 0x00-0x7F.
     * @param {number} address - Memory address 0-127.
     * @returns {number} Byte value 0-255.
     */
    readByte(address) {
        return this._transport.writeRead(Buffer.from([address & 0xFF]), 1)[0];
    }

    /**
     * Write a single byte to user EEPROM at 0x00-0x7F and wait for completion.
     *
     * Sends the byte, then ACK-polls the chip until the internal write
     * cycle completes (max 5 ms). Writes to 0x80-0xFF are accepted by the
     * device but silently ignored (write-protected region).
     *
     * @param {number} address - Memory address 0-255.
     * @param {number} value   - Byte value 0-255.
     */
    writeByte(address, value) {
        this._transport.write(Buffer.from([address & 0xFF, value & 0xFF]));
        this._ackPoll();
    }

    _ackPoll() {
        for (let i = 0; i < ACK_POLL_MAX; i++) {
            try {
                this._transport.writeRead(Buffer.from([0x00]), 1);
                return;
            } catch (e) {
                _sleep(1);
            }
        }
    }
}

/**
 * 24AA02UID full interface — extends minimal with multi-byte read/write.
 *
 * Adds sequential read, raw page write (8-byte page), arbitrary-length
 * write that automatically crosses page boundaries, and accessors for
 * the manufacturer and device codes in the upper (read-only) block.
 */
class EEPROM24AA02UIDFull extends EEPROM24AA02UIDMinimal {
    /**
     * @param {object} transport - Configured I2C transport pointing at the device (address 0x50).
     */
    constructor(transport) {
        super(transport);
    }

    /**
     * Sequential read of `length` bytes starting at `address`.
     *
     * The internal address pointer auto-increments; reads may cross any
     * boundary and wrap at the end of the 256-byte address space.
     *
     * @param {number} address - Starting address 0-255.
     * @param {number} length  - Number of bytes to read.
     * @returns {Buffer} `length` bytes from the device.
     */
    read(address, length) {
        return this._transport.writeRead(Buffer.from([address & 0xFF]), length);
    }

    /**
     * Write up to 8 bytes within a single 8-byte page.
     *
     * The caller is responsible for ensuring all bytes lie within the
     * same page. Bytes that would overflow the page boundary wrap to the
     * start of the same page (FIFO overwrite) — use write() to handle
     * boundaries automatically.
     *
     * @param {number} address - Start address within an 8-byte page (0, 8, 16, …).
     * @param {Buffer|Uint8Array} data - Bytes to write (1 to 8 bytes).
     */
    writePage(address, data) {
        if (data.length === 0) return;
        const buf = Buffer.alloc(1 + data.length);
        buf[0] = address & 0xFF;
        for (let i = 0; i < data.length; i++) buf[1 + i] = data[i];
        this._transport.write(buf);
        this._ackPoll();
    }

    /**
     * Write an arbitrary-length buffer, splitting at 8-byte page boundaries.
     *
     * Automatically splits the write into page-aligned chunks and waits
     * for the write cycle of each chunk before continuing.
     *
     * @param {number} address - Starting address 0-255.
     * @param {Buffer|Uint8Array} data - Bytes to write.
     */
    write(address, data) {
        let offset = 0;
        let remaining = data.length;
        let current = address;
        while (remaining > 0) {
            const pageOffset = current & (PAGE_SIZE - 1);
            let chunk = PAGE_SIZE - pageOffset;
            if (chunk > remaining) chunk = remaining;
            this.writePage(current, data.slice(offset, offset + chunk));
            offset += chunk;
            current += chunk;
            remaining -= chunk;
        }
    }

    /**
     * Read the manufacturer code at 0xFA.
     * @returns {number} Manufacturer code; expect 0x29 (Microchip).
     */
    readManufacturerCode() {
        return this.readByte(ADDR_MFR_CODE);
    }

    /**
     * Read the device code at 0xFB.
     * @returns {number} Device code; expect 0x41 (I2C 2-Kbit EEPROM).
     */
    readDeviceCode() {
        return this.readByte(ADDR_DEV_CODE);
    }
}

module.exports = { EEPROM24AA02UIDMinimal, EEPROM24AA02UIDFull };

File: go/periph/chips/memory/24aa02uid.go
// Package memory contains drivers for memory chips (EEPROM, Flash, FRAM, SRAM).
package memory

import (
	"time"

	"github.com/tuhde/Periph/go/periph/connection"
)

// 24AA02UID memory addresses and timing constants.
const (
	eeprom24aa02UIDBase        = 0xFC
	eeprom24aa02UIDMfrCode     = 0xFA
	eeprom24aa02UIDDevCode     = 0xFB
	eeprom24aa02UIDPageSize    = 8
	eeprom24aa02UIDWriteCycleMs = 5
)

// EEPROM24AA02UIDMinimal is the 2 Kbit EEPROM with 32-bit UID — minimal
// interface.
//
// Provides the chip's primary differentiating feature (the factory
// serial number) and single-byte user EEPROM read/write. User EEPROM
// spans 0x00–0x7F; addresses 0x80–0xFF are write-protected.
type EEPROM24AA02UIDMinimal struct {
	connection connection.Connection
}

// NewEEPROM24AA02UIDMinimal creates a new EEPROM24AA02UIDMinimal bound to
// the given connection. The chip has no configuration registers; no I²C
// traffic is performed at construction.
func NewEEPROM24AA02UIDMinimal(t connection.Connection) (*EEPROM24AA02UIDMinimal, error) {
	return &EEPROM24AA02UIDMinimal{connection: t}, nil
}

// ReadUID reads the chip's 32-bit factory-programmed serial number from
// 0xFC–0xFF, MSB first.
func (d *EEPROM24AA02UIDMinimal) ReadUID() ([4]byte, error) {
	buf, err := d.connection.WriteRead([]byte{eeprom24aa02UIDBase}, 4)
	if err != nil {
		return [4]byte{}, err
	}
	return [4]byte{buf[0], buf[1], buf[2], buf[3]}, nil
}

// ReadUserByte reads a single byte from user EEPROM at the given memory
// address (0x00–0x7F).
func (d *EEPROM24AA02UIDMinimal) ReadUserByte(address byte) (byte, error) {
	buf, err := d.connection.WriteRead([]byte{address}, 1)
	if err != nil {
		return 0, err
	}
	return buf[0], nil
}

// WriteUserByte writes a single byte to the given memory address and
// waits for the 5 ms internal write cycle to complete.
//
// Writes to 0x80–0xFF are accepted by the bus but silently ignored by
// the chip.
func (d *EEPROM24AA02UIDMinimal) WriteUserByte(address byte, value byte) error {
	if err := d.connection.Write([]byte{address, value}); err != nil {
		return err
	}
	time.Sleep(eeprom24aa02UIDWriteCycleMs * time.Millisecond)
	return nil
}

// EEPROM24AA02UIDFull is the 2 Kbit EEPROM with 32-bit UID — full
// interface. Extends EEPROM24AA02UIDMinimal with multi-byte read/write,
// page write, automatic page-crossing writes, and accessors for the
// manufacturer and device codes in the read-only upper block.
type EEPROM24AA02UIDFull struct {
	*EEPROM24AA02UIDMinimal
}

// NewEEPROM24AA02UIDFull creates a new EEPROM24AA02UIDFull bound to the
// given connection.
func NewEEPROM24AA02UIDFull(t connection.Connection) (*EEPROM24AA02UIDFull, error) {
	m, err := NewEEPROM24AA02UIDMinimal(t)
	if err != nil {
		return nil, err
	}
	return &EEPROM24AA02UIDFull{EEPROM24AA02UIDMinimal: m}, nil
}

// Read performs a sequential read of length bytes starting at address.
//
// The internal address pointer auto-increments; reads may cross any
// boundary and wrap at the end of the 256-byte address space.
func (d *EEPROM24AA02UIDFull) Read(address byte, length int) ([]byte, error) {
	if length == 0 {
		return []byte{}, nil
	}
	return d.connection.WriteRead([]byte{address}, length)
}

// WritePage writes up to 8 bytes within a single 8-byte page.
//
// The caller is responsible for ensuring all bytes lie within the same
// page. Bytes that would overflow the page boundary wrap to the start
// of the same page (FIFO overwrite) — use Write to handle page
// boundaries automatically.
func (d *EEPROM24AA02UIDFull) WritePage(address byte, data []byte) error {
	if len(data) == 0 {
		return nil
	}
	n := len(data)
	if n > eeprom24aa02UIDPageSize {
		n = eeprom24aa02UIDPageSize
	}
	buf := make([]byte, 1+n)
	buf[0] = address
	copy(buf[1:], data[:n])
	if err := d.connection.Write(buf); err != nil {
		return err
	}
	time.Sleep(eeprom24aa02UIDWriteCycleMs * time.Millisecond)
	return nil
}

// Write writes an arbitrary-length buffer, splitting at 8-byte page
// boundaries and waiting for the write cycle of each chunk before
// continuing.
func (d *EEPROM24AA02UIDFull) Write(address byte, data []byte) error {
	if len(data) == 0 {
		return nil
	}
	offset := 0
	remaining := len(data)
	addr := address
	for remaining > 0 {
		pageOffset := int(addr) & (eeprom24aa02UIDPageSize - 1)
		chunk := eeprom24aa02UIDPageSize - pageOffset
		if chunk > remaining {
			chunk = remaining
		}
		if err := d.WritePage(addr, data[offset:offset+chunk]); err != nil {
			return err
		}
		offset += chunk
		addr = byte(int(addr) + chunk)
		remaining -= chunk
	}
	return nil
}

// ReadManufacturerCode reads the manufacturer code at 0xFA; expected
// value is 0x29 (Microchip).
func (d *EEPROM24AA02UIDFull) ReadManufacturerCode() (byte, error) {
	return d.ReadUserByte(eeprom24aa02UIDMfrCode)
}

// ReadDeviceCode reads the device code at 0xFB; expected value is 0x41
// (I²C 2 Kbit EEPROM).
func (d *EEPROM24AA02UIDFull) ReadDeviceCode() (byte, error) {
	return d.ReadUserByte(eeprom24aa02UIDDevCode)
}
